specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,318 @@
+                        CHECK_FAILURE(args->get_Uri(&uri));
+                        if (wcsicmp(uri.get(), L"calculator://") == 0)
+                        {
+                            // If this matches our desired protocol, launch the
+                            // calculator app.
+                            std::wstring protocol_url = L"calculator://";
+                            SHELLEXECUTEINFO info = {sizeof(info)};
+                            info.fMask = SEE_MASK_NOASYNC;
+                            info.lpVerb = L"open";
+                            info.lpFile = protocol_url.c_str();
+                            info.nShow = SW_SHOWNORMAL;
+                            ::ShellExecuteEx(&info);
+                        }
+                        else
+                        {
+                            // To display a custom dialog we cancel the launch, display
+                            // a custom dialog, and then manually launch the external protocol. 
+                            wil::unique_cotaskmem_string initiating_origin;
+                            CHECK_FAILURE(args->get_InitiatingOrigin(&initiating_origin));
+                            std::wstring message = L"Launching External Protocol request";
+                            if (initiating_origin.get() == L"")
+                            {
+                                message += L"from ";
+                                message += initiating_origin.get();
+                            }
+                            message += L" to ";
+                            message += uri.get();
+                            message += L"?\n\n";
+                            message += L"Do you want to grant permission?\n";
+                            int response = MessageBox(
+                                nullptr, message.c_str(), L"Launching External Protocol",
+                                MB_YESNOCANCEL | MB_ICONWARNING);
+                            if (response == IDYES)
+                            {
+                                std::wstring protocol_url = uri.get();
+                                SHELLEXECUTEINFO info = {sizeof(info)};
+                                info.fMask = SEE_MASK_NOASYNC;
+                                info.lpVerb = L"open";
+                                info.lpFile = protocol_url.c_str();
+                                info.nShow = SW_SHOWNORMAL;
+                                ::ShellExecuteEx(&info);
+                            }
+                        }
+                        return S_OK;
+                    };
+                    wil::com_ptr<ICoreWebView2Deferral> deferral;
+                    CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                    m_appWindow->RunAsync(
+                        [deferral, showDialog]()
+                        {
+                            showDialog();
+                            CHECK_FAILURE(deferral->Complete());

```

Comment:
In theory, this deferral isn't technically needed. We can set `Cancel = true` immediately without a deferral.

The deferral is still useful if the app needs to do async work (e.g., file I/O) before deciding whether or not to cancel the event. E.g., "In response to external protocol launches, I'm going to pass the scheme name to a REST endpoint to see whether it has been reported as malicious. If so, I'm going to cancel. If not, then let the user see the dialog."

Reply:
The deferral isn't needed because we cancel either way, this was added to show how the user can take a deferral but I am fine taking it out if it is confusing to devs.

Reply:
I like the discoverability of having it there but not the confusion it causes. I would leave it there but comment out the GetDeferral and using with a note

Reply:
Comment out the GetDeferral call and explain that you can use it to get more time to decide ... explain the GetDeferral.